library/src/updater.rs

```diff
@@ -161,22 +178,53 @@ pub fn init(
         return Err(InitError::AlreadyInitialized);
     }
 
-    let _ = with_config(|config| {
-        UpdaterState::load_or_new_on_error(
+    handle_prior_boot_failure_if_necessary()
+}
+
+/// If, at initialization time, we detect that we were in the process of booting a patch, report a
+/// failure to boot for that patch and queue an event to report the failure.
+pub fn handle_prior_boot_failure_if_necessary() -> Result<(), InitError> {
+    with_config(|config| {
+        let mut state = UpdaterState::load_or_new_on_error(
             &config.storage_dir,
             &config.release_version,
             config.patch_public_key.as_deref(),
-        )
-        .on_init()
-    });
+        );
+        if let Some(patch) = state.currently_booting_patch() {
+            state.record_boot_failure_for_patch(patch.number)?;
+            state.queue_event(patch_event(
+                config,
+                EventType::PatchInstallFailure,
+                patch.number,
+            ))?;
+        }
 
-    Ok(())
+        Ok(())
+    })
+    .map_err(|e| {
+        error!("Failed to clean up after a failed patch: {:?}", e);
+        InitError::FailedToCleanUpFailedPatch
+    })
 }
 
+/// Whether the auto-update flag is set to true in the config.
 pub fn should_auto_update() -> anyhow::Result<bool> {
     with_config(|config| Ok(config.auto_update))
 }
 
+/// Creates a `PatchEvent` for the given `EventType` and patch number for reporting to the server.
+fn patch_event(config: &UpdateConfig, event_type: EventType, patch_number: usize) -> PatchEvent {
```

Comment:
Is this essentially a constructor/factory?  Does it need to be closer to PatchEvent?

Reply:
Yeah, good call. I was mimicking the `patch_check_request` function below, but it definitely makes more sense as a constructor. Done.